builder/builder-next/control/control.go

```diff
@@ -256,9 +275,30 @@ func (c *Controller) Solve(ctx context.Context, req *controlapi.SolveRequest) (*
 		return nil, err
 	}
 	if req.Exporter != "" {
-		exp, err := w.Exporter(req.Exporter, c.opt.SessionManager)
-		if err != nil {
-			return nil, err
+		var exp exporter.Exporter
+		if c.opt.UseSnapshotter {
+			if req.Exporter == mobyexporter.Moby {
+				req.Exporter = client.ExporterImage
+			}
+
+			exp, err = w.Exporter(req.Exporter, c.opt.SessionManager)
```

Comment:
Possibly silly question (trying to get my head around all the wrapping that we do);

- We have a `ContainerdWorker`, which we instantiate if we create a `newSnapshotterController` and a `newGraphDriverController`
- We get the right exporter to use from the worker that we use


With both of the above; would `worker.Exporter` not be able to return the right exporter, and that exporter be smart enough to perform the validation / conversion of `req.ExporterAttrs["name"]` to the right format, and to set the `unpack` option?

Reply:
I don't think we should do this.

Here is what we have:

- the `ConatinerdWorker` returns the right exporter (switches to buildkit's image exporter if the client asked for a `moby` exporter
- the control.go controls/configures the exporter
- the exporter's job is to export the image

If we do what you are asking we will have to create our own exporter that will be a copy/paste of the buildkit's containerimage exporter. So there is no real gain and we might lose the link between a control and a worker since we will add additional logic to the worker and make it less configurable.

Note that some if not all of the control code will go away once we update buildkit since in latest buildkit the containerimage exporter does unpack by default https://github.com/moby/buildkit/blob/abcab1426b1c2a2201c4e5665f99f399273047a9/exporter/containerimage/export.go#LL125C3-L125C3

Reply:
> If we do what you are asking we will have to create our own exporter that will be a copy/paste of the buildkit's containerimage exporter. So there is no real gain and we might lose the link between a control and a worker since we will add additional logic to the worker and make it less configurable.

I was thinking along these lines; have the magic exporter attributes handled by the exporter (which is a thin wrapper around the upstream which is missing some of the things);

```go
package snapshotexporter

import (
	"context"
	"strings"

	"github.com/moby/buildkit/exporter"
)

type snapshotExporter struct {
	exp exporter.Exporter
}

func New(exp exporter.Exporter) (exporter.Exporter, error) {
	return &snapshotExporter{exp: exp}, nil
}

func (e *snapshotExporter) Resolve(ctx context.Context, ExporterAttrs map[string]string) (exporter.ExporterInstance, error) {
	reposAndTags, err := SanitizeRepoAndTags(strings.Split(ExporterAttrs["name"], ","))
	if err != nil {
		return nil, err
	}
	ExporterAttrs["name"] = strings.Join(reposAndTags, ",")
	ExporterAttrs["unpack"] = "true"
	return e.exp.Resolve(ctx, ExporterAttrs)
}
```

Then have the Worker wrap the BuildKit exporter;


```go
func (w *ContainerdWorker) Exporter(name string, sm *session.Manager) (exporter.Exporter, error) {
	switch name {
	case mobyexporter.Moby, client.ExporterImage:
		exp, err := w.Worker.Exporter(client.ExporterImage, sm)
		if err != nil {
			return nil, err
		}
		return snapshotexporter.New(exp)
	default:
		return w.Worker.Exporter(name, sm)
	}
}
```

Reply:
DONE